### PR TITLE
Dedupe proxy rapids shuffle manager byte code

### PR DESCRIPTION
**Pull request by gerashegalov**
Stop inheriting ShuffleManaager in base class. Spark ShuffleManager causes bytecode discrepancies in the unshimmed area. Inherit it only in the version-specific code:

Previously, only 3.1+ was bitwise-identical
```
$ grep ProxyRapidsShuffleInternalManagerBase.class dist/target/binary-diffs/*
dist/target/binary-diffs/spark311cdh-spark312.identical:org/apache/spark/sql/rapids/ProxyRapidsShuffleInternalManagerBase.class
dist/target/binary-diffs/spark311cdh-spark320.identical:org/apache/spark/sql/rapids/ProxyRapidsShuffleInternalManagerBase.class
```

After the PR, bitwise-identical  across all supported 3.x 
```
$ grep ProxyRapidsShuffleInternalManagerBase.class dist/target/binary-diffs/*
dist/target/binary-diffs/spark311cdh-spark302.identical:org/apache/spark/sql/rapids/ProxyRapidsShuffleInternalManagerBase.class
dist/target/binary-diffs/spark311cdh-spark312.identical:org/apache/spark/sql/rapids/ProxyRapidsShuffleInternalManagerBase.class
dist/target/binary-diffs/spark311cdh-spark320.identical:org/apache/spark/sql/rapids/ProxyRapidsShuffleInternalManagerBase.class
```

Other changes:
- Shim Dev doc
- Fixes incorrect path for `org.apache.spark.sql.rapids.shims.spark301.RapidsShuffleInternalManager` 
    
Signed-off-by: Gera Shegalov <gera@apache.org>

### CODE REVIEW

**Review by revans2 — COMMENTED** [review]
Some comments explaining what is happening would be nice. It is a little confusing who calls into whom and what the different levels of indirection are for.

**Comment by gerashegalov** [comment]
> Some comments explaining what is happening would be nice. It is a little confusing who calls into whom and what the different levels of indirection are for.  

@revans2 I agree. However, this was a subtle area where I found it safer to make the Proxy to follow the existing implementation  in lockstep. I added a shim README to explain the general background. I think we can get rid of some of the intermediate classes. I prefer it to be done on 21.12 branch

**Comment by gerashegalov** [comment]
build

File: shims/README.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+Apache Spark such as 3.1.x, 3.2.x, and a number of vendor releases that contain
+a mix of patches from different upstream releases. These artifacts are generally 
+incompatible between each other, at both source code level and even more often 
+at the binary level. The role of the Shim layer is to hide these issues from the 
+common code, maximize reuse, and minimize logic duplication. 
+
+This is achieved by using a ServiceProvider pattern. All Shims implement the same API, 
+the suitable Shim implementation is loaded after detecting the current Spark build version
+attempting to instantiate our plugin. We use the
+[ShimLoader](https://github.com/NVIDIA/spark-rapids/blob/branch-21.10/sql-plugin/src/main/scala/com/nvidia/spark/rapids/ShimLoader.scala)
```
Comment (pxLi) [inline shims/README.md:19]: nit: link to relative path

File: shims/README.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+however it does not work easily with
+[Scala packages](https://contributors.scala-lang.org/t/scala-signature-layout/3327/4)
+
+So instead we resort to the idea of JDK's ParallelWorldClassloader in combination with the fact that
+Spark runtime uses mutable classloaders we can alter after detecting the runtime version.
+Using JarURLConnection URLs we create a Parallel World of the current version within the jar, e.g.:
+
+Spark 3.0.2's URLs:
+```
+jar:file:/home/spark/rapids-4-spark_2.12-21.10.jar!/
```
Comment (pxLi) [inline shims/README.md:65]: nit: would be nice to have 21.10.**0** w/ patch version

File: shims/README.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+The third issue may arise from the fact that the shared logic may transitively reference a class that 
+for one another reason resides in a Parallel World. Untangling this is tedious and may be unnecessary.
+The following approach robustly prevents from running into issues related to that. 
+
+We know that at the time such a class is loaded by Spark it's not strictly needed if the Plugin 
+has not been loaded yet. More accurately, it may not be strictly needed until later when the first 
+query can be run when the Spark SQL session and its extensions are initialized. It improves the 
+user experience if the first query is not penalized beyond necessary though. By design, Plugin guarantees 
+that the classloader is 
+[set up at load time](https://github.com/NVIDIA/spark-rapids/blob/branch-21.10/sql-plugin/src/main/scala/com/nvidia/spark/SQLPlugin.scala#L29)
```
Comment (pxLi) [inline shims/README.md:108]: nit: relative path

**Comment by gerashegalov** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala [hunk 4]
```diff
@@ -419,33 +419,47 @@ abstract class RapidsShuffleInternalManagerBase(conf: SparkConf, val isDriver: B
   }
 }
 
+/**
+ * Trait that makes it easy to check whether we are dealing with the
+ * a RAPIDS Shuffle Manager
+ *
+ * TODO name does not match its function anymore
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala:426]: Are you going to fix this TODO?

**Review by revans2 — APPROVED** [review]

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala (hunk 4 above)
Comment (gerashegalov) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala:426]: I filed https://github.com/NVIDIA/spark-rapids/issues/3624 for this.

**Review by abellina — APPROVED** [review]
Thanks for the thorough README change, that was useful. Could use a follow on issue for `VisibleShuffleManager`?

**Pull request merged**